lib/core/src/Cardano/Wallet/Api/Link.hs

```diff
@@ -0,0 +1,520 @@
+import Cardano.Wallet.Primitive.Types
+    ( AddressState, Hash, PoolId, SortOrder, WalletId )
+import Data.Function
+    ( (&) )
+import Data.Generics.Internal.VL.Lens
+    ( (^.) )
+import Data.Generics.Product.Typed
+    ( HasType, typed )
+import Data.Maybe
+    ( isJust )
+import Data.Proxy
+    ( Proxy (..) )
+import Data.Text
+    ( Text )
+import GHC.TypeLits
+    ( KnownSymbol, sameSymbol )
+import Network.HTTP.Types.Method
+    ( Method )
+import Servant.API
+    ( (:>)
+    , Capture
+    , Header'
+    , IsElem
+    , QueryParam
+    , ReflectMethod (..)
+    , ReqBody
+    , StdMethod (..)
+    , Verb
+    )
+import Servant.API.TypeLevel
+    ( Elem )
+import Servant.Links
+    ( HasLink (..), safeLink' )
+import Web.HttpApiData
+    ( ToHttpApiData (..) )
+
+import qualified Cardano.Wallet.Api as Api
+
+--
+-- Wallets
+--
+
+-- NOTE
+-- Type-class is necessary here because to type-check 'IsElem  endpoint Api'
+-- below. We can't actually prove that 'PostByronWallet style' is present in the
+-- API for any 'style' (counter example: 'Shelley').
+class PostWallet (style :: WalletStyle) where
+    postWallet :: (Method, Text)
+
+instance PostWallet 'Shelley where
+    postWallet = endpoint @Api.PostWallet id
+
+instance KnownSymbol (WalletStyleSymbol 'Random) => PostWallet 'Random where
+    postWallet = endpoint @(Api.PostByronWallet 'Random) id
+
+instance KnownSymbol (WalletStyleSymbol 'Icarus) => PostWallet 'Icarus where
+    postWallet = endpoint @(Api.PostByronWallet 'Icarus) id
+
+instance KnownSymbol (WalletStyleSymbol 'Trezor) => PostWallet 'Trezor where
+    postWallet = endpoint @(Api.PostByronWallet 'Trezor) id
+
+instance KnownSymbol (WalletStyleSymbol 'Ledger) => PostWallet 'Ledger where
+    postWallet = endpoint @(Api.PostByronWallet 'Ledger) id
+
+deleteWallet
+    :: forall (style :: WalletStyle) w.
+        ( HasType (ApiT WalletId) w
+        , KnownSymbol (WalletStyleSymbol style)
+        )
+    => w
+    -> (Method, Text)
+deleteWallet w = discriminate @style
+    (endpoint @Api.DeleteWallet (wid &))
+    (endpoint @Api.DeleteByronWallet (wid &))
+  where
+    wid = w ^. typed @(ApiT WalletId)
+
+getWallet
+    :: forall (style :: WalletStyle) w.
+        ( HasType (ApiT WalletId) w
+        , KnownSymbol (WalletStyleSymbol style)
+        )
+    => w
+    -> (Method, Text)
+getWallet w = discriminate @style
+    (endpoint @Api.GetWallet (wid &))
+    (endpoint @Api.GetByronWallet (wid &))
+  where
+    wid = w ^. typed @(ApiT WalletId)
+
+getUTxOsStatistics
+    :: forall (style :: WalletStyle) w.
+        ( HasType (ApiT WalletId) w
+        , style ~ 'Shelley
+        )
+    => w
+    -> (Method, Text)
+getUTxOsStatistics w =
+    endpoint @Api.GetUTxOsStatistics (wid &)
+  where
+    wid = w ^. typed @(ApiT WalletId)
+
+listWallet
+    :: forall (style :: WalletStyle).
+        ( KnownSymbol (WalletStyleSymbol style)
+        )
+    => (Method, Text)
+listWallet = discriminate @style
```

Comment:
I would suggest renaming this wrapper function to `listWallets` (for consistency).